### ISSUE TITLE
[Develop] Scheduler only runs jobs one time

### ISSUE THREAD

**Issue by gtmanfred** [saltstack/salt#40022]
### Description of Issue/Question
The schedule only runs the jobs on the startup of the salt master.  I have a suspicions that this also happens on the minion, but haven't checked.  It runs once when the schedule gets started, and doesn't run after that.  For some runners, it doesn't run at all.

Reverting #39527 remedies the problem

### Setup
```
schedule:
  job1:
    function: test.stdout_print
    minutes: 1
```

Start the master and look for lines like this 

```
2017-03-14 19:37:18,895 [salt.utils.schedule                      ][INFO    ][4985] Running scheduled job: job1
2017-03-14 19:37:18,895 [salt.utils.schedule                      ][DEBUG   ][4985] schedule: This job was scheduled with jid_include, adding to cache (jid_include defaults to True)
2017-03-14 19:37:18,896 [salt.utils.schedule                      ][DEBUG   ][4985] schedule: This job was scheduled with a max number of 1
```

### Versions Report

```
[root@ip-10-0-0-120 salt]# salt --versions-report
Salt Version:
           Salt: 2016.11.3-3263-g64b9d5a

Dependency Versions:
           cffi: 1.9.1
       cherrypy: Not Installed
       dateutil: 2.6.0
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: 2.17
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.5 (default, Nov  6 2016, 00:28:07)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: centos 7.2.1511 Core
        machine: x86_64
        release: 3.10.0-327.10.1.el7.x86_64
         system: Linux
        version: CentOS Linux 7.2.1511 Core
```

**Comment by gtmanfred** [comment]
@YuriGrigorov you might be interested in this.

**Comment by YuriGrigorov** [comment]
Thanks for your report.

The problem is in master design. We can see in `run` function in `salt/master.py` that scheduled jobs (`self.handle_schedule`) are evaluated with other maintenance stuff:
```
while True:
    now = int(time.time())
    if (now - last) >= self.loop_interval:
        salt.daemons.masterapi.clean_old_jobs(self.opts)
        salt.daemons.masterapi.clean_expired_tokens(self.opts)
    self.handle_search(now, last)
    self.handle_git_pillar()
    self.handle_schedule()
    self.handle_key_cache()
    self.handle_presence(old_present)
    self.handle_key_rotate(now)
    salt.daemons.masterapi.fileserver_update(self.fileserver)
    salt.utils.verify.check_max_open_files(self.opts)
    last = now
    time.sleep(self.loop_interval)
```
so whie default `loop_interval` is 60 seconds, interval between scheduler evaluations is about 61.82 and we are 1.82 secs runout. In my patch i added check only for 1 second.
```
if -1 < seconds <= 0:
    run = True
```
To fix this problem we have some options:
* log message that evaluation interval is more than job interval and you need to decrease loop_interval on increase job interval
* increase check for example till 3 seconds `if -3 < seconds <= 0`:
* or we can run job every time when we are running out of interval `if seconds <= 0`:

**Comment by gtmanfred** [comment]
Is there a reason to not set it to `if seconds <= 0` ?

**Comment by YuriGrigorov** [comment]
If you configure job with interval of 90 seconds it will wire after 120. And i think it's not what you want.

**Comment by YuriGrigorov** [comment]
We should add some ``grace_time`` -- seconds after the designated runtime that the job is still allowed to be run. And if it's exceed then reschedule the job.

**Comment by thatch45** [comment]
The solution is to check the time between evaluations and sleep for the remaining time. So you do a time.time.() at the top of the schedule eval and then again right before the sleep, take the extra time and subtract it from the loop interval and sleep that long, and if the subtracted number is negative then skip the sleep.

**Comment by thatch45** [comment]
https://github.com/saltstack/salt/blob/develop/salt/thorium/__init__.py#L166

Is where we do it in Thorium...

**Comment by YuriGrigorov** [comment]
Please take a look at https://github.com/YuriGrigorov/salt/commit/1257d64dfa68f2889f3a1784eea3452751431ad4

**Comment by thatch45** [comment]
Looks good, I made one comment.

**Comment by YuriGrigorov** [comment]
I prepared another version. I think it's a little bit better. Take a look please.

**Comment by gtmanfred** [comment]
If you open a PR I think that would be a better way to discuss this.

Thanks,
Daniel

**Comment by YuriGrigorov** [comment]
Ok. First I'll make some test and then open PR.

**Comment by cachedout** [comment]
Closed via #40103
